benchmarks/cudf-merge.py

```diff
@@ -361,10 +361,10 @@ def main():
     print_multi(values=["Row matching fraction", f"{args.frac_match}"])
     print_separator(separator="=", length=110)
     print_multi(values=["Wall-clock", f"{format_time(wc)}"])
-    print_multi(values=["Bandwidth", f"{format_bytes(bw)}/s"])
-    print_multi(values=["Throughput", f"{format_bytes(tp)}/s"])
+    print_multi(values=["Avg Bandwidth / GPU", f"{format_bytes(bw)}/s"])
+    print_multi(values=["Throughput / GPU", f"{format_bytes(tp)}/s"])
     print_separator(separator="=", length=110)
-    print_multi(values=["Run", "Wall-clock", "Bandwidth", "Throughput"])
+    print_multi(values=["Run", "Wall-clock", "Avg Bandwidth / GPU", "Throughput / GPU"])
```

Comment:
Sorry, I screwed up. I hadn't spotted that `n_chunks` is multiplied by the total number of GPUs, so these are aggregate rather than per-GPU numbers.

Reply:
Right, but the throughput is a single worker multiplied by `n_chunks`, i.e., only that worker's throughput. So that's not an aggregate of the full cluster?

Reply:
`parse_args` does:

```
    args.n_chunks = len(args.devs) * args.chunks_per_dev
```

So I think that this measures the total data (certainly if I run the code, the
data volume is consistent with more GPUs -> higher throughput), no?

Reply:
I think this is the final issue with this change for me. Due to the way `args.n_chunks` is updated in `parse_args`, the printouts here are really the aggregate bandwidth (not per GPU):

```
$ python cudf-merge.py --warmup-iter 2 --frac-match 0.3 -c 10_000_000 -d 0,1 --chunks-per-dev 1 --iter 10 
cuDF merge benchmark
--------------------------------------------------------------------------------------------------------------
Device(s)                 | [0, 1]
Chunks per device         | 1
Rows per chunk            | 10000000
Total data processed      | 5.96 GiB
Data processed per iter   | 610.35 MiB
Row matching fraction     | 0.3
==============================================================================================================
Wall-clock                | 391.33 ms
Avg Bandwidth / GPU       | 13.71 GiB/s
Throughput / GPU          | 15.23 GiB/s
==============================================================================================================
Run                       | Wall-clock                | Avg Bandwidth / GPU       | Throughput / GPU
0                         | 38.73 ms                  | 14.04 GiB/s               | 15.39 GiB/s
1                         | 40.32 ms                  | 12.46 GiB/s               | 14.78 GiB/s
2                         | 38.30 ms                  | 13.90 GiB/s               | 15.56 GiB/s
3                         | 39.93 ms                  | 14.14 GiB/s               | 14.93 GiB/s
4                         | 37.84 ms                  | 13.49 GiB/s               | 15.75 GiB/s
5                         | 38.26 ms                  | 13.92 GiB/s               | 15.58 GiB/s
6                         | 38.08 ms                  | 13.83 GiB/s               | 15.65 GiB/s
7                         | 38.00 ms                  | 13.85 GiB/s               | 15.69 GiB/s
8                         | 37.89 ms                  | 13.93 GiB/s               | 15.73 GiB/s
9                         | 38.15 ms                  | 13.74 GiB/s               | 15.63 GiB/s
```
vs.
```
$ python cudf-merge.py --warmup-iter 2 --frac-match 0.3 -c 10_000_000 -d 0,1,2,3 --chunks-per-dev 1 --iter 10 
cuDF merge benchmark
--------------------------------------------------------------------------------------------------------------
Device(s)                 | [0, 1, 2, 3]
Chunks per device         | 1
Rows per chunk            | 10000000
Total data processed      | 11.92 GiB
Data processed per iter   | 1.19 GiB
Row matching fraction     | 0.3
==============================================================================================================
Wall-clock                | 391.67 ms
Avg Bandwidth / GPU       | 24.57 GiB/s
Throughput / GPU          | 30.44 GiB/s
==============================================================================================================
Run                       | Wall-clock                | Avg Bandwidth / GPU       | Throughput / GPU
0                         | 36.86 ms                  | 27.46 GiB/s               | 32.34 GiB/s
1                         | 39.28 ms                  | 25.39 GiB/s               | 30.35 GiB/s
2                         | 37.59 ms                  | 23.88 GiB/s               | 31.71 GiB/s
3                         | 39.75 ms                  | 22.00 GiB/s               | 29.99 GiB/s
4                         | 37.14 ms                  | 25.90 GiB/s               | 32.10 GiB/s
5                         | 37.70 ms                  | 25.11 GiB/s               | 31.62 GiB/s
6                         | 36.74 ms                  | 26.41 GiB/s               | 32.44 GiB/s
7                         | 37.22 ms                  | 25.15 GiB/s               | 32.02 GiB/s
8                         | 39.57 ms                  | 22.13 GiB/s               | 30.13 GiB/s
9                         | 43.67 ms                  | 23.50 GiB/s               | 27.30 GiB/s
```

So I think just removing the ` / GPU` is correct.

Reply:
I'm normally skeptical of comparing different GPUs in a DGX-1 because of the different interconnects they have available. Part of the interconnects are NV1 (25GB/s) and the remaining are NV2 (50GB/s) as seen with `nvidia-smi topo -m`, and given GPUs `0,1` are NV1, that is at least partially responsible for the bandwidth increase.

A DGX-2 is more reliable in that matter, as bandwidths should be almost equal due to NVSwitch. In this case we still see `0,1,2,3` having higher average bandwidth than `0,1`, but not as significant as in DGX-1 which showed pretty much linear scaling. See below:

```
$ python cudf-merge.py --warmup-iter 2 --frac-match 0.3 -c 10_000_000 -d 0,1 --chunks-per-dev 1 --iter 10
cuDF merge benchmark
--------------------------------------------------------------------------------------------------------------
Device(s)                 | [0, 1]
Chunks per device         | 1
Rows per chunk            | 10000000
Total data processed      | 5.96 GiB
Data processed per iter   | 610.35 MiB
Row matching fraction     | 0.3
==============================================================================================================
Wall-clock                | 280.93 ms
Avg Bandwidth / GPU       | 31.95 GiB/s
Throughput / GPU          | 21.22 GiB/s
==============================================================================================================
Run                       | Wall-clock                | Avg Bandwidth / GPU       | Throughput / GPU
0                         | 31.78 ms                  | 20.59 GiB/s               | 18.75 GiB/s
1                         | 27.74 ms                  | 33.78 GiB/s               | 21.48 GiB/s
2                         | 27.58 ms                  | 33.25 GiB/s               | 21.61 GiB/s
3                         | 26.92 ms                  | 35.33 GiB/s               | 22.14 GiB/s
4                         | 27.00 ms                  | 35.76 GiB/s               | 22.07 GiB/s
5                         | 27.22 ms                  | 32.73 GiB/s               | 21.89 GiB/s
6                         | 26.70 ms                  | 35.41 GiB/s               | 22.33 GiB/s
7                         | 27.10 ms                  | 34.20 GiB/s               | 22.00 GiB/s
8                         | 27.41 ms                  | 32.37 GiB/s               | 21.74 GiB/s
9                         | 26.92 ms                  | 33.85 GiB/s               | 22.14 GiB/s

$ python cudf-merge.py --warmup-iter 2 --frac-match 0.3 -c 10_000_000 -d 0,1,2,3 --chunks-per-dev 1 --iter 10
cuDF merge benchmark
--------------------------------------------------------------------------------------------------------------
Device(s)                 | [0, 1, 2, 3]
Chunks per device         | 1
Rows per chunk            | 10000000
Total data processed      | 11.92 GiB
Data processed per iter   | 1.19 GiB
Row matching fraction     | 0.3
==============================================================================================================
Wall-clock                | 300.60 ms
Avg Bandwidth / GPU       | 42.57 GiB/s
Throughput / GPU          | 39.66 GiB/s
==============================================================================================================
Run                       | Wall-clock                | Avg Bandwidth / GPU       | Throughput / GPU
0                         | 29.62 ms                  | 45.72 GiB/s               | 40.24 GiB/s
1                         | 29.15 ms                  | 47.19 GiB/s               | 40.90 GiB/s
2                         | 29.37 ms                  | 44.94 GiB/s               | 40.60 GiB/s
3                         | 29.07 ms                  | 46.73 GiB/s               | 41.00 GiB/s
4                         | 28.88 ms                  | 48.47 GiB/s               | 41.28 GiB/s
5                         | 31.43 ms                  | 34.46 GiB/s               | 37.93 GiB/s
6                         | 31.60 ms                  | 32.29 GiB/s               | 37.72 GiB/s
7                         | 28.41 ms                  | 48.06 GiB/s               | 41.96 GiB/s
8                         | 28.27 ms                  | 49.47 GiB/s               | 42.16 GiB/s
9                         | 30.27 ms                  | 37.62 GiB/s               | 39.38 GiB/s
```

For a DGX-1, I think a more fair comparison would be to set `UCX_TLS=^cuda_ipc` to only use InfiniBand, providing equal bandwidth between all endpoints, in which case we get:

```
$ UCX_TLS=^cuda_ipc python cudf-merge.py --warmup-iter 2 --frac-match 0.3 -c 10_000_000 -d 0,1 --chunks-per-dev 1 --iter 10
cuDF merge benchmark
--------------------------------------------------------------------------------------------------------------
Device(s)                 | [0, 1]
Chunks per device         | 1
Rows per chunk            | 10000000
Total data processed      | 5.96 GiB
Data processed per iter   | 610.35 MiB
Row matching fraction     | 0.3
==============================================================================================================
Wall-clock                | 696.12 ms
Avg Bandwidth / GPU       | 4.88 GiB/s
Throughput / GPU          | 8.56 GiB/s
==============================================================================================================
Run                       | Wall-clock                | Avg Bandwidth / GPU       | Throughput / GPU
0                         | 68.34 ms                  | 4.93 GiB/s                | 8.72 GiB/s
1                         | 69.05 ms                  | 4.89 GiB/s                | 8.63 GiB/s
2                         | 68.63 ms                  | 4.89 GiB/s                | 8.69 GiB/s
3                         | 68.08 ms                  | 4.95 GiB/s                | 8.76 GiB/s
4                         | 71.25 ms                  | 4.87 GiB/s                | 8.37 GiB/s
5                         | 68.88 ms                  | 4.87 GiB/s                | 8.65 GiB/s
6                         | 68.80 ms                  | 4.88 GiB/s                | 8.66 GiB/s
7                         | 69.08 ms                  | 4.84 GiB/s                | 8.63 GiB/s
8                         | 69.02 ms                  | 4.85 GiB/s                | 8.64 GiB/s
9                         | 69.26 ms                  | 4.84 GiB/s                | 8.61 GiB/s

$ UCX_TLS=^cuda_ipc python cudf-merge.py --warmup-iter 2 --frac-match 0.3 -c 10_000_000 -d 0,1,2,3 --chunks-per-dev 1 --iter 10
cuDF merge benchmark
--------------------------------------------------------------------------------------------------------------
Device(s)                 | [0, 1, 2, 3]
Chunks per device         | 1
Rows per chunk            | 10000000
Total data processed      | 11.92 GiB
Data processed per iter   | 1.19 GiB
Row matching fraction     | 0.3
==============================================================================================================
Wall-clock                | 919.79 ms
Avg Bandwidth / GPU       | 5.03 GiB/s
Throughput / GPU          | 12.96 GiB/s
==============================================================================================================
Run                       | Wall-clock                | Avg Bandwidth / GPU       | Throughput / GPU
0                         | 91.15 ms                  | 5.00 GiB/s                | 13.08 GiB/s
1                         | 91.24 ms                  | 5.04 GiB/s                | 13.07 GiB/s
2                         | 91.64 ms                  | 5.02 GiB/s                | 13.01 GiB/s
3                         | 91.26 ms                  | 5.05 GiB/s                | 13.06 GiB/s
4                         | 91.15 ms                  | 5.05 GiB/s                | 13.08 GiB/s
5                         | 91.94 ms                  | 5.02 GiB/s                | 12.97 GiB/s
6                         | 91.45 ms                  | 5.02 GiB/s                | 13.04 GiB/s
7                         | 91.11 ms                  | 5.06 GiB/s                | 13.08 GiB/s
8                         | 91.00 ms                  | 5.06 GiB/s                | 13.10 GiB/s
9                         | 91.53 ms                  | 5.00 GiB/s                | 13.02 GiB/s
```

In the above case results are not as significant as well, just like the DGX-2. In other words, I'm not sure if this alone proves/disproves if we should write `/ GPU` or not. What do you think @wence- ?

Reply:
The throughput/GPU should be at best constant (since the partition by hash and local merge are perfectly data parallel, whereas the the communication only costs something that increases with the number of GPUs). So I don't know how the throughput/GPU  can almost double going from 2 to 4 GPUs.

Reply:
On the DGX-2 case for example, the wall-clock is just slightly lower (~10%) for 2 GPUs when compared to the 4 GPUs sample, even though we increased the problem size with more GPUs. That would cause throughput/GPU to go up, given the problem size is now larger, no?

Reply:
> On the DGX-2 case for example, the wall-clock is just slightly lower (~10%) for 2 GPUs when compared to the 4 GPUs sample, even though we increased the problem size with more GPUs. That would cause throughput/GPU to go up, given the problem size is now larger, no?

I don't think so. Here's my explanation:

On 2 GPUs, the benchmark processes a total of 615MB data in 70 ms (or about 8.8GB/s). So the processing rate per GPU is 4.4 GB/s;

On 4 GPUs, the benchmark processes a total of 1.2GB data in 92 ms (or about 13GB/s). So the processing rate per GPU is 3.25 GB/s.

The _overall_ throughput is higher (as we would hope). But for throughput/GPU to go up, the wallclock would have to decrease as we add more GPUs: we're doing a weak scaling experiment here where we are increasing the amount of data to be processed linearly and linearly adding more compute to deal with it.

Reply:
Yeah, ok, based on the total processed data I agree that your explanation looks correct. I reverted that change in https://github.com/rapidsai/ucx-py/pull/867/commits/ae884a6dbfaa282ba27d2a6d3a419d5025b3dd5d .

However, in the future we may want to refactor a bit pieces of this code to make sure that it's more immediately evident. I think the fact that we are discussing whether this is right or not feels like something could be improved. For now, let's leave it as is.